libs/cardano-ledger-core/src/Cardano/Ledger/SafeHash.hs

```diff
@@ -1,3 +1,7 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
```

Comment:
How come this is necessary? Is it because of the weird thing with ghc8 vs 9?

Reply:
This is necessary because `HashAnnotated` in `indexProxy` in this module is indeed redundant. As far as GHC is concerned it is not needed. However, I kept it there to improve backward compatibility, because previously it was part of that type class and usage of `indexProxy` would have resulted in requiring this constraint. Also if I'd remove the constraint, then type inference would suffer, because without `HashAnnotated` the type of `i` would not get automatically resolved from `x`, but with functional dependency in `HashAnnotated` it will get resolved. 

All of this isn't necessary at all, because I could not find any usage of `indexProxy` in ledger or downstream. But, it doesn't hurt to do be diligent even if it is not used :slightly_smiling_face: